### PR TITLE
fix: chart state and series functions cleanup

### PR DESCRIPTION
**Pull request by monfera**
## Summary

Mild refactoring that leaves logic and types in place: 
- fixes the handling of zero non-global chart types
- fixes a sorting issue with non-finite returns
- removes some branching and dries up conditionals
- removes avoidable variables, `let`s and conditional mutation
- attempts to improve the readability and "flatness" of some logic, converting some triangular bits to linear and rectangular shape
- regularizes `return`s into one or just very few, easy to spot places
- removes some type related branching that used to (or should have, sans TS limitations) narrowed to a singular type or `never`
- added a heterogeneous small multiples test case we looked at together
- removes IDE-generated JSDocs `@param`s encountered

No functional or API change. Precedes further work on non-cartesian small multiples.

### Checklist

Delete any items that are not applicable to this PR.

### CODE REVIEW

File: src/chart_types/xy_chart/utils/series.ts [hunk 1]
```diff
@@ -557,41 +544,28 @@ export function getSeriesName(
...
-    }
-  } else {
-    name = nameKeys.join(delimiter);
-  }
-
-  return name;
+  const multipleYAccessors = spec && spec.yAccessors.length > 1;
+  const nameKeys = multipleYAccessors ? seriesIdentifier.seriesKeys : seriesIdentifier.seriesKeys.slice(0, -1);
+  const nonZeroLength = nameKeys.length > 0;
+
```
Comment (monfera) [inline src/chart_types/xy_chart/utils/series.ts:561]: The block below (`return nonZeroLength...`) could be extracted into a separate function, though not much benefit right now

File: src/chart_types/xy_chart/utils/series.ts [hunk 2]
```diff
@@ -518,6 +503,8 @@ function getSortedOrdinalXValues(
   }
 }
 
+const BIG_NUMBER = Number.MAX_SAFE_INTEGER; // the sort comparator must yield finite results, can't use infinities
```
Comment (markov00) [inline src/chart_types/xy_chart/utils/series.ts:506]: 😬

File: stories/small_multiples/6_heterogeneous_cartesians.tsx [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    >
+      MIN
+    </span>
+  );
+  const showLegend = boolean('Show Legend', false);
+  const onElementClick = action('onElementClick');
+
+  return (
+    <Chart className="story-chart">
+      <Settings onElementClick={onElementClick} rotation={90} showLegend={showLegend} />
```
Comment (markov00) [inline stories/small_multiples/6_heterogeneous_cartesians.tsx:79]: Let's rotate this horizontally to avoid vertical lines and areas

**Review by markov00 — APPROVED** [review]
All good for me, thanks for taking time to clean this up, in particular the state reducer,
I've just set a comment to rotate the added chart horizontally that looks a bit better IMHO

File: src/chart_types/xy_chart/utils/series.ts (hunk 2 above)
Comment (monfera) [reply src/chart_types/xy_chart/utils/series.ts:506]: not very pretty, just safer than the previous `Infinity` 😸

File: stories/small_multiples/6_heterogeneous_cartesians.tsx (hunk 3 above)
Comment (monfera) [reply stories/small_multiples/6_heterogeneous_cartesians.tsx:79]: 👍

File: stories/small_multiples/6_heterogeneous_cartesians.tsx [hunk 4]
```diff
@@ -0,0 +1,164 @@
+/*
```
Comment (nickofthyme) [inline stories/small_multiples/6_heterogeneous_cartesians.tsx:1]: 👍🏼

File: src/utils/common.ts [hunk 5]
```diff
@@ -539,3 +539,5 @@ export const getPercentageValue = <T>(ratio: string | number, relativeValue: num
 
   return num && !isNaN(num) ? num : defaultValue;
 };
+
+export const keepDistinct = <T>(d: T, i: number, a: T[]): boolean => a.indexOf(d) === i;
```
Comment (nickofthyme) [inline src/utils/common.ts:543]: Can you add a description to this helper function?

**Review by nickofthyme — CHANGES_REQUESTED** [review]
I don't agree with __all__ the if statements being converted to nested ternaries, in some cases they a cleaner but in many cases, the compact logic is hard to follow. I think this should be a preference thing when using nesting ternaries is required.

File: src/chart_types/xy_chart/utils/series.ts [hunk 6]
```diff
@@ -557,41 +544,28 @@ export function getSeriesName(
...
-
-    if (customLabel !== null) {
-      return customLabel.toString();
-    }
+  const customLabel =
+    typeof spec?.name === 'function'
+      ? spec.name(seriesIdentifier, isTooltip)
+      : typeof spec?.name === 'object' // extract booleans once https://github.com/microsoft/TypeScript/issues/12184 is fixed
+      ? getSeriesNameFromOptions(spec.name, seriesIdentifier, spec.name.delimiter ?? SERIES_DELIMITER)
+      : null;
```
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/xy_chart/utils/series.ts:552]: I think it's not nesting b/c it's a linear fallthrough case, ie. no structural nesting applies, more like a `case` sequence in a `switch`. Would work with indented formatting, would still preserve ability to linearly read through

File: src/state/chart_state.ts [hunk 7]
```diff
@@ -295,52 +295,29 @@ export const chartStoreReducer = (chartId: string) => {
...
-          const internalChartState = initInternalChartState(chartType);
-          return {
-            ...state,
-            specsInitialized: true,
-            specParsing: false,
-            chartType,
-            internalChartState,
-          };
-        }
+        const chartType = chartTypeFromSpecs(state.specs);
```
[the remark this answers is not in the document]
Comment (monfera) [reply src/state/chart_state.ts:298]: The PR description and my code notes lists such items, and may be easier to see by commit, but the best might be a call to discuss. Glad to write details if you prefer that

File: src/utils/common.ts (hunk 5 above)
Comment (monfera) [reply src/utils/common.ts:543]: 👍

File: src/state/chart_state.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (nickofthyme) [reply src/state/chart_state.ts:298]: A call on Monday would be fine with me.

File: src/state/actions/colors.ts [hunk 8]
```diff
@@ -29,36 +29,36 @@ export const SET_TEMPORARY_COLOR = 'SET_TEMPORARY_COLOR';
...
   key: SeriesKey;
   color: Color | null;
 }
 
-interface SetPersistedColor {
+interface PersistedColor {
   type: typeof SET_PERSISTED_COLOR;
   key: SeriesKey;
   color: Color | null;
 }
```
[the remark this answers is not in the document]
Comment (monfera) [reply src/state/actions/colors.ts:46]: Good point about consistency, glad to revert for now, will need to catch up on this convention

File: src/chart_types/xy_chart/utils/series.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (nickofthyme) [reply src/chart_types/xy_chart/utils/series.ts:552]: It's a ternary inside another ternary, that's nested no?

As far as the indenting, this is what I'd like to see that delineates the `exprIfTrue` and `exprIfTrue` code blocks.

```ts
const customLabel =
    typeof spec?.name === 'function'
      ? spec.name(seriesIdentifier, isTooltip)
      : typeof spec?.name === 'object' // extract booleans once https://github.com/microsoft/TypeScript/issues/12184 is fixed
        ? getSeriesNameFromOptions(spec.name, seriesIdentifier, spec.name.delimiter ?? SERIES_DELIMITER)
        : null;
```

File: src/chart_types/xy_chart/utils/series.ts [hunk 9]
```diff
@@ -19,20 +19,20 @@
...
 import { Logger } from '../../../utils/logger';
 import { ColorConfig } from '../../../utils/themes/theme';
 import { groupSeriesByYGroup, isHistogramEnabled, isStackedSpec } from '../domains/y_domain';
 import { LastValues } from '../state/utils/types';
 import { applyFitFunctionToDataSeries } from './fit_function_utils';
 import { groupBy } from './group_data_series';
-import { BasicSeriesSpec, SeriesTypes, SeriesSpecs, SeriesNameConfigOptions, StackMode } from './specs';
-import { formatStackedDataSeriesValues, datumXSortPredicate } from './stacked_series_utils';
+import { BasicSeriesSpec, SeriesNameConfigOptions, SeriesSpecs, SeriesTypes, StackMode } from './specs';
+import { datumXSortPredicate, formatStackedDataSeriesValues } from './stacked_series_utils';
```
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/xy_chart/utils/series.ts:35]: Not sure how to interpret this or what my takeaway should be, as even now, sorting is mandatory, as in:
- I can't commit a file if imports are not in the order that the linter wants
- these reorders were not done by me manually, it's automatic reordering
- if you think the problem is the noise in the commit: yes it's a bit annoying, but it happened precisely because of that rule, or not? You could've written, this is where avoiding [this and this rule] might help as then it wouldn't introduce not very interesting ordering

In short, pls add what the lesson learnt is in your view

File: stories/small_multiples/6_heterogeneous_cartesians.tsx (hunk 3 above)
Comment (monfera) [reply stories/small_multiples/6_heterogeneous_cartesians.tsx:79]: 3545cd83118bfc0103e91da8c167555bd8b60313

File: src/chart_types/xy_chart/utils/series.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (nickofthyme) [reply src/chart_types/xy_chart/utils/series.ts:35]: Oh I thought these were manually done by you. Disregard

File: src/chart_types/xy_chart/utils/series.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/xy_chart/utils/series.ts:552]: This formatting would be OK too, maybe in some formatting PR we can test drive how the rule would shape the ternaries here and elsewhere.
You are right, the syntax AST is a [right-deep tree](https://d3i71xaburhd42.cloudfront.net/c551ceaf1b7f8460af2c186bdaec0126864d31fa/2-Figure1-1.png) which is conceptually still linear (the reason it was born), as in, you could substitute it with a `reduce`, or `switch/case` alternative, or it can be read as 

```
Condition1
Result 1
Condition2
Result2
Condition3
Result3
...
```
Especially with code I didn't write, I find it way easier to linearly step through, "OK this condition still not true, let's look whether the next applies" compared to needing to simulate ascending and descending into (semantically, not just AST-wise) nested `if`s. So this is what I meant by rectangular code, linear readability. I know I'm in trouble when the left edge of the code is meandering, I might have some weaknesses there, usually, it takes less time to simplify data flow and control flow paths and making the code shorter (all of which I think also happened) than to eyeball it long enough to grok what's going on. 

Languages:
JS doesn't offer a true `cond`, and the equivalent `switch(true)` [pattern](https://medium.com/trabe/using-switch-true-in-javascript-986e8ad8ae4f) is not only a yoda condition but also a bit unwieldy (b/c our linter requires curlies for even single-row returns), though often better than nested `if`. A `reduce` with the options would work too, but it'd eagerly evaluate never taken options, expensive and would need a lot of defensive optional chaining etc.
The best support in this case is provided by those that have pattern matching constructs. Eg, a Fibonacci in haskell would be
```haskell
fib n | n == 0 = 1
      | n == 1 = 1
      | n >= 2 = fib (n-1) + fib (n-2)
```
again, linear/rectangular code, easy to read top to bottom, and a single pass reading gives a good picture, no need to ascend and descend into conditional valleys. JS don't have no such pattern matching...

Bigger picture:
Both code (internal view, our DX) and API (external view, Kibana DX) benefit from a strive toward simplicity. One big way for achieving it is to reduce the number of conditional paths, nesting variety and one-off handling. It's always straightforward to add to the code; "on this condition X let's do something different before/after/instead of the current thing", which is how code bases become untangled. 

So my feel is that the real solution is _not_ to turn into ternary all such functions, though it has some positives too here, and also, _not_ to introduce code level constraints. 

Instead, we could be be a bit slower to add features; let's brainstorm if 
- we actually need a feature (eg. this function is made this complex in part by the `null` return of the callback we discussed today; meanwhile nobody uses it, and even our test cases don't lock in the promising yet speculative flexibility)
- can it be achieved without added complexity, or how can we minimize added complexity
- can we remove something while solving a need by going a bit more general

That's the drawback with broadening types into union types (discriminated or not), and/or optional, and the error prone falsey and fallthrough cases that require mental juggling and exponentially growing conditional branches like `| null | undefined`. This is why I feel it was good to chat about lending the user overriding and other config options in ways that retain as much homogeneity as possible, and type clarity helps the user too, fewer types, tighter doc etc. This is all I'm learning along the way so quote me on this when running into my subpar code, past or future 😸 

So ultimately I agree, and work toward not needing this many layers of "pattern matching", ternaries or not. Future PRs from all of us FTW!

File: src/state/chart_state.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (monfera) [reply src/state/chart_state.ts:298]: 👍 although the partition small multiples are based on this so I'd be keen to possibly get approval today, I could merge tomorrow as a Monday afternoon (my time) chat would be a bit of a blocker. We could put our head together and further tweak in follow-up PRs, I hope this already has a number of tangible improvements and yes a couple of contentional formatting points 😄

File: src/utils/common.ts (hunk 5 above)
Comment (monfera) [reply src/utils/common.ts:547]: 17e135643294236c5af444bc0155ea5fa7567d6c

File: src/state/actions/colors.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (monfera) [reply src/state/actions/colors.ts]: 17e135643294236c5af444bc0155ea5fa7567d6c

File: src/state/chart_state.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (monfera) [reply src/state/chart_state.ts:298]: Just took another look at the file you ask about, a list is:
- fixed the issue that a chart with zero non-global chartTypes didn't raise a warning before, and returned an `undefined` that's not allowed by the function return type
- simplified the `findMainChartType` code  
- simplified a couple of types eg. removes an `as` assertion
- collapsed the occasional branching returns into single returns

The commit names (see the first several commits) and the file compare are pretty self-expanatory, so I think maybe what's left for Monday is any concern about any of these (if any), and maybe the code branchiness topic

**Comment by monfera** [comment]
Re-titled to `fix` as it fixes two low likelihood runtime issues (zero chartTypes -> undefined return; Infinity and NaNs in sort function) and improves types (fewer asserts, removal of a never taken code path)

File: src/chart_types/xy_chart/utils/series.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (nickofthyme) [reply src/chart_types/xy_chart/utils/series.ts:552]: Thanks as always for the thoughtful reply. I just don't want to get to a point where we are forming super complex ternaries to remove complex nested conditionals. Replacing some makes sense while others would be hard to understand, but yes reducing the bushyness would be great.

As for the `null` option and other non-used functionality. I think we could go through the code sometime and maybe identify features in the code that are any of the following:

- Not used and does not have a practicable use case.
- Features that are not used but have a good use case. (documentation is to blame)
- Features or options that have multiple (>2) options to use but only a simple and complex option would suffice.

**Review by nickofthyme — APPROVED** [review]
Latest changes LGTM. Thanks for the comments.

**Comment by nickofthyme** [comment]
@monfera I think `fix` is also good, I think a more concise title is fine. Feel free to change it back if you like.

File: src/chart_types/xy_chart/utils/series.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/xy_chart/utils/series.ts:552]: Thanks @nickofthyme!

Great idea for hunting for disused functionality (esp. that's not foreseen to be used; maybe this `null` or some alternative would have great use). Maybe we could look out for the replaceability of optional or nullable/undefinable properties too, at least for alternatives discussion. Though both take a good bunch of time.

I agree the ternary chaining is less than optimal, there's room for improving on it, hopefully by simplification rather than eg. `if/return` or `let/if/assign` chaining or nesting the next time one of us touches that part of the code, so it's likely not for eternity.
